ingest-asset-reconciler/src/main/scala/uk/gov/nationalarchives/Lambda.scala

```diff
@@ -60,62 +60,73 @@ class Lambda extends LambdaRunner[Input, StateOutput, Config, Dependencies] {
-                    val assetChildTitle = if (titleOrFileName.isEmpty) fileNameWithoutExtension else titleOrFileName
-                    titleOfCoWithoutExtension == assetChildTitle
+            urlsToIoRepresentations <- dependencies.entityClient.getUrlsToIoRepresentations(entity.ref, Some(representationType))
+            contentObjects <- urlsToIoRepresentations.map { urlToIoRepresentation =>
+              val generationVersion = urlToIoRepresentation.reverse.takeWhile(_ != '/').toInt
+              dependencies.entityClient.getContentObjectsFromRepresentation(entity.ref, representationType, generationVersion)
+            }.flatSequence
+
+            _ <- log("Content Objects, belonging to the representation, have been retrieved from API")
+
+            stateOutput <-
+              if (contentObjects.isEmpty)
+                IO.pure(
+                  StateOutput(wasReconciled = false, s"There were no Content Objects returned for entity ref '${entity.ref}'")
+                )
+              else {
+                for {
+                  bitstreamInfoPerContentObject <- contentObjects
+                    .map(co => dependencies.entityClient.getBitstreamInfo(co.ref))
+                    .flatSequence
+
+                  _ <- log(s"Bitstreams of Content Objects have been retrieved from API")
+                } yield {
+                  val childrenThatDidNotMatchOnChecksum =
+                    childrenForRepresentationType.filter { assetChild =>
+                      val bitstreamWithSameChecksum = bitstreamInfoPerContentObject.find { bitstreamInfoForCo =>
+                        assetChild.checksumSha256 == bitstreamInfoForCo.fixity.value &&
+                        bitstreamInfoForCo.potentialCoTitle.exists { titleOfCo => // DDB titles don't have file extensions, CO titles do
+                          val titleOfCoWithoutExtension = titleOfCo.split('.').dropRight(1).mkString(".")
+                          lazy val fileNameWithoutExtension = assetChild.name.split('.').dropRight(1).mkString(".")
+                          val titleOrFileName = assetChild.title.getOrElse(fileNameWithoutExtension)
+                          val assetChildTitle = if (titleOrFileName.isEmpty) fileNameWithoutExtension else titleOrFileName
+                          titleOfCoWithoutExtension == assetChildTitle
+                        }
+                      }
+
+                      bitstreamWithSameChecksum.isEmpty
+                    }
+
+                  if (childrenThatDidNotMatchOnChecksum.isEmpty) StateOutput(wasReconciled = true, "")
+                  else {
+                    val idsOfChildrenThatDidNotMatchOnChecksum = childrenThatDidNotMatchOnChecksum.map(_.id)
+                    StateOutput(
+                      wasReconciled = false,
+                      s"Out of the ${childrenForRepresentationType.length} files expected to be ingested for assetId '${input.assetId}' with representationType $representationType, " +
```

Comment:
If we do what I suggested above, then maybe we should add the suffix to this message too?

Reply:
As above, the error message as is is helpful enough not to need the extra code.